#11 - create+switch default branch on release

Ocramius opened the pull request:
Fixes #11 by providing default logic for switching to a new default branch at
every release. Specifically:

 * when a newer release target branch exists, use that (i.e. release `1.2.3` => existing branch `1.5.x`,
   switch to `1.5.x`)
 * when a newer release branch does not exist, create it (i.e. release `1.2.3` => existing branch `1.2.x`,
   create `1.3.x` from `1.2.x`, use `1.3.x`)
 * when no release branches exist, skip entire execution

TODOs:

 * [x] try this out in a sandbox repository
 * [x] better documentation? (new `feature/` files)
 * [x] update examples
 * [x] get rid of special handling for `master`, which is not interesting to us anymore
 * [x] document token requirements

Note: it would be good to get this into `1.0.0`, since it breaks BC by changing semantics around the `master` branch naming.

Ocramius committed:
Initial draft for #11 - create+switch default branch on release

Fixes #11 by providing default logic for switching to a new default branch at
every release. Specifically:

 * when a newer release target branch exists, use that (i.e. release `1.2.3` => existing branch `1.5.x`,
   switch to `1.5.x`)
 * when a newer release branch does not exist, create it (i.e. release `1.2.3` => existing branch `1.2.x`,
   create `1.3.x` from `1.2.x`, use `1.3.x`)
 * when no release branches exist, skip entire execution

TODOs:

 * [ ] try this out in a sandbox repository
 * [ ] better documentation?
 * [ ] update examples

Ocramius commented on src/Git/Value/MergeTargetCandidateBranches.php:
```diff
@@ -101,4 +102,41 @@ public function branchToMergeUp(SemVerVersion $version): ?BranchName
             : $branch;
     }
+
+    /** @TODO move `master` exclusion into ctor - `master` is no longer a valid release branch */
```
This change will likely lead to a BC break. Not a problem, since the library is not yet released, but good to keep in mind that the tool will change massively, and that `feature/` files will need adjusting

Ocramius commented on src/Git/Value/MergeTargetCandidateBranches.php:
```diff
@@ -101,4 +102,41 @@ public function branchToMergeUp(SemVerVersion $version): ?BranchName
+        ))[0] ?? null;
+    }
+
+    /** @TODO move `master` exclusion into ctor - `master` is no longer a valid release branch */
```
Same as https://github.com/laminas/automatic-releases/pull/12/files#r456971170

Ocramius commented:
Trying this out in the real world, it seems that my github actions `GITHUB_TOKEN` doesn't suffice for this sort of interaction, since that token cannot be used to `PATCH /repos/:org/:name`

Also tried to use a custom token, but no luck: pinging github support.

Ocramius commented:
Yep, seems like `GITHUB_TOKEN` is scoped too strictly for this: `PATCH /repos/:org/:name` is not possible with that, and instead a custom admin-alike token is needed.

Don't know yet what the precise scope for it is.

Using a custom token produced the expected results at https://github.com/Ocramius/just-trying-automated-releases/runs/888025697?check_suite_focus=true, in which, by releasing `1.1.0`, branch `1.2.x` was created, published, and the repository switched to it.

Ocramius commented:
Confirmed with github support:

 > The `GITHUB_TOKEN` is limited these permissions in the repository containing the workflow:
 > 
 > https://docs.github.com/en/actions/configuring-and-managing-workflows/authenticating-with-the-github_token#permissions-for-the-github_token
 > 
 > The following page details which GitHub App permissions are required for each endpoint and shows that making changes to the repository's settings, including changing the default branch, requires `write` on `administration`:
 > 
 > https://docs.github.com/en/rest/reference/permissions-required-for-github-apps#permission-on-administration
 > 
 > If you create a personal access token, the token will need the `repo` permission and will need to belong to a user with `admin` permissions on the repository.

rieschl commented on src/Application/Command/SwitchDefaultBranchToNextMinor.php:
```diff
@@ -0,0 +1,82 @@
+        if ($newestBranch === null) {
+            $output->writeln('No stable branches found: cannot switch default branch');
+
+            return 0;
```
shouldn't that return a non-zero code? also, there are `Command::SUCCESS` and `Command::FAILURE` for better readability.

Ocramius replied:
This is not intended to make the build fail: it is true that no stable branches may exist at all (happens when tagging a first release)

rieschl commented on src/Github/Api/V3/SetDefaultBranchThroughApiCall.php:
```diff
@@ -0,0 +1,69 @@
+    ): void {
+        $request = $this->messageFactory
+            ->createRequest(
+                'PATCH',
```
`\Fig\Http\Message\RequestMethodInterface::METHOD_PATCH`?

Ocramius replied:
Good one! I didn't even know of these constants \o/

Ocramius replied:
Hmm, no, those interfaces don't exist in current dependencies :\

rieschl replied:
it's here https://github.com/php-fig/http-message-util/blob/master/src/RequestMethodInterface.php i.e. in `fig/http-message-util`

Ocramius replied:
Yeh, I would avoid importing a new package just for that

Ocramius commented on examples/.github/workflows/release-on-milestone-closed.yml:
```diff
@@ -35,3 +35,13 @@ jobs:
+        with:
+          command-name: "laminas:automatic-releases:switch-default-branch-to-next-minor"
+        env:
+          "GITHUB_TOKEN": ${{ secrets.GITHUB_TOKEN }}
```
As discussed in https://github.com/laminas/automatic-releases/pull/12#issuecomment-660946161, this will need a separate token

Ocramius replied:
This now uses a completely independent environment variable

Ocramius committed:
#11 described branch switching behavior with practical examples

Ocramius committed:
#11 ensuring that scenarios completely ignore `master` and `develop` branches in their behavior expectations

Ocramius committed:
#11 removed special rules around the handling of `master`, which is no longer considered the "next major" branch

This used to be an old convention of `doctrine/*` projects, and it is also gone from `doctrine/*`,
where the naming will likely be adjusted to make branch target versions more clear to contributors.

Ocramius committed:
Fix #11 documented the necessity of having an admin token for switching default branches

weierophinney commented on feature/automated-releases.feature:
```diff
@@ -23,39 +27,56 @@ Feature: Automated releases
   Scenario: If a major release branch exists, the tool creates a major release from there
     Given following existing branches:
-      | name   |
-      | 1.0.x  |
-      | master |
+      | name    |
+      | 1.0.x   |
+      | 2.0.x   |
+      | master  |
+      | develop |
     And following open milestones:
       | name  |
       | 2.0.0 |
     When I close milestone "2.0.0"
     Then tag "2.0.0" should have been created on branch "master"
     And branch "2.0.x" should have been created from "master"
```
This scenario confuses me:

- Why would the 2.0.0 tag be created from the `master` branch?
- Why would it create the `2.0.x` branch from `master`, considering the branch already exists?

Ocramius replied:
This is indeed an oversight: I will adjust 👍

Ocramius replied:
Adjusted: also, no branch created here: that's done by the `laminas:automatic-releases:switch-default-branch-to-next-minor` command instead.

weierophinney commented on feature/automated-releases.feature:
```diff
@@ -23,39 +27,56 @@ Feature: Automated releases
-      | name   |
-      | 1.0.x  |
-      | master |
+      | name    |
+      | 1.0.x   |
+      | 2.0.x   |
+      | master  |
+      | develop |
     And following open milestones:
       | name  |
       | 1.1.0 |
     When I close milestone "1.1.0"
-    Then tag "1.1.0" should have been created on branch "master"
-    And branch "1.1.x" should have been created from "master"
+    Then the tool should have halted with an error
```
I'm assuming this is because the `1.1.x` branch does not exist? (just want a clarification)

Ocramius replied:
Correct: no auto-magic branch switching 👍

weierophinney commented on feature/automated-releases.feature:
```diff
@@ -23,39 +27,56 @@ Feature: Automated releases
+      | develop |
     And following open milestones:
       | name  |
       | 1.1.0 |
     When I close milestone "1.1.0"
-    Then tag "1.1.0" should have been created on branch "master"
-    And branch "1.1.x" should have been created from "master"
+    Then the tool should have halted with an error
 
   Scenario: If a minor release branch exists, the tool creates a new minor release from there
     Given following existing branches:
       | name   |
       | 1.1.x  |
+      | 1.2.x  |
       | master |
     And following open milestones:
       | name  |
       | 1.1.0 |
     When I close milestone "1.1.0"
     Then tag "1.1.0" should have been created on branch "1.1.x"
-    And a new pull request from branch "1.1.x" to "master" should have been created
+    And a new pull request from branch "1.1.x" to "1.2.x" should have been created
```
This one feels mis-named. Maybe "If a minor release branch exists, when closing the minor release milestone, the tool tags the minor release from the branch, and creates a pull request against the next newer minor release branch."

Ocramius replied:
It is correct for the tool fo have created a PR from 1.1.x to 1.2.x here though: are you suggesting a change in wording?

Ocramius replied:
Got it now: wording adjusted

weierophinney commented on feature/automated-releases.feature:
```diff
@@ -23,39 +27,56 @@ Feature: Automated releases
-    And a new pull request from branch "1.1.x" to "master" should have been created
+    And a new pull request from branch "1.1.x" to "1.2.x" should have been created
+
+  Scenario: If a no newer release branch exists, the tool will not create a pull request against it
```
```suggestion
  Scenario: If no newer release branch exists, the tool will not create any pull requests
```

weierophinney commented on feature/default-branch-switching.feature:
```diff
@@ -0,0 +1,85 @@
+      | 1.3.x  |
+    And the default branch should be "1.3.x"
+
+  Scenario: A pre-existing minor release branch is set as default branch on release
```
```suggestion
  Scenario: The latest pre-existing minor release branch is set as default branch on a successful release
```

weierophinney reviewed (changes requested):
Mainly a few questions.

Ocramius committed:
Adjust wording around the expectation of no PRs created

Co-authored-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

Ocramius committed:
Better wording around expected default branch switch scenario

Co-authored-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

asgrim commented on src/Github/Api/V3/SetDefaultBranchThroughApiCall.php:
```diff
@@ -0,0 +1,69 @@
+            )
+            ->withAddedHeader('Content-Type', 'application/json')
+            ->withAddedHeader('User-Agent', 'Ocramius\'s minimal API V3 client')
+            ->withAddedHeader('Authorization', 'bearer ' . $this->apiToken);
```
Is `Bearer` case sensitive?

Ocramius replied:
Apparently not: tried it manually though

asgrim commented on README.md:
```diff
@@ -25,6 +25,11 @@ to your project or organisation:
+   then you have to provide a `ORGANIZATION_ADMIN_TOKEN`, which is a github token with administrative rights over
+   your organisation. This is required for the `laminas:automatic-releases:switch-default-branch-to-next-minor`
+   command, because [changing default branch of a repository currently requires administrative token rights](https://developer.github.com/v3/repos/#update-a-repository).
+   You can generate a token from your [personal access tokens page](https://github.com/settings/tokens/new).
```
The "type" of token is a personal access token, but the naming suggests a different "type" (an "organization admin token"). The link here is the only clue it's a personal access token issued to a user who has organization admin rights. Not sure how to clarify this, but it took me a bit of working out reading the docs given.

Ocramius replied:
Added further detail to this wording

asgrim commented on src/Application/Command/SwitchDefaultBranchToNextMinor.php:
```diff
@@ -0,0 +1,82 @@
+        $event          = $this->loadGithubEvent->__invoke();
+        $repositoryPath = $this->variables->githubWorkspacePath();
+
+        Assert::directory($repositoryPath . '/.git');
```
Does `symfony/console` catch exceptions and turn them into non-zero exit codes? Or do we need special handling if this assertion fails?

Ocramius replied:
Yes, exceptions are caught and turned into non-zero exit codes :+1:

asgrim commented on src/Git/Value/BranchName.php:
```diff
@@ -45,11 +45,6 @@ public function isReleaseBranch(): bool
         return preg_match('/^(v)?\d+\\.\d+(\\.x)?$/', $this->name) === 1;
     }
 
-    public function isNextMajor(): bool
```
Not `private`, is this a missed BC break? `roave/backward-compatibility-check` could help identify stuff like this, as we already reached `1.*`

Ocramius replied:
Luckily, this library has `"type": "project"`, so it's not meant to be incorporated and relied upon at code level.

What we need to keep stable is the CLI behavior :+1:

Ocramius replied:
> we already reached `1.*`

We're working on `1.0.x`, but we never released BTW

asgrim commented on src/Git/Value/BranchName.php:
```diff
@@ -70,6 +65,13 @@ public function majorAndMinor(): array
         return [$major, $minor];
     }
 
+    public function targetVersion(): SemVerVersion
```
Not a fan of this generic naming; my question is "what is the target?" surely the target varies depending on the type of release being done? Just a minor naming thing, admittedly.

Ocramius replied:
Reworded to clarify what this `SemVerVersion` is :)

Ocramius committed:
#11 s/targetVersion/targetMinorReleaseVersion (better method clarity)

Ref: https://github.com/laminas/automatic-releases/pull/12#discussion_r461193914

Ocramius committed:
#11 clarifying that `ORGANIZATION_ADMIN_TOKEN` comes from an admin user issuing a personal token with admin scope

Ocramius committed:
#11 clarified on leftover `master` branch reference: tool will pick major release branch for new major releases

Ocramius committed:
#11 clarifying merge-up PR scenario description

Ref: https://github.com/laminas/automatic-releases/pull/12#discussion_r461142193
